operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java

```diff
@@ -0,0 +1,138 @@
+            waitForKcToFail(kc);
+        } catch (Exception e) {
+            savePodLogs();
+            throw e;
+        }
+    }
+
+    @Test
+    public void testDeploymentDurability() throws InterruptedException {
+        try {
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            var deploymentName = kc.getMetadata().getName();
+            deployKeycloak(k8sclient, kc, true);
+
+            Log.info("Trying to delete deployment");
+            assertThat(k8sclient.apps().deployments().withName(deploymentName).delete()).isTrue();
+            Awaitility.await()
+                    .untilAsserted(() -> assertThat(k8sclient.apps().deployments().withName(deploymentName).get()).isNotNull());
+
+            waitForKeycloakToBeReady(k8sclient, kc); // wait for reconciler to calm down to avoid race condititon
+
+            Log.info("Trying to modify deployment");
+
+            var deployment = k8sclient.apps().deployments().withName(deploymentName).get();
+            var labels = Map.of("address", "EvergreenTerrace742");
+            var flandersEnvVar = new EnvVarBuilder().withName("NEIGHBOR").withValue("Stupid Flanders!").build();
+            var origSpecs = deployment.getSpec();
+
+            deployment.getMetadata().getLabels().putAll(labels);
+            deployment.getSpec().getTemplate().getSpec().getContainers().get(0).setEnv(List.of(flandersEnvVar));
+            k8sclient.apps().deployments().createOrReplace(deployment);
+
+            Awaitility.await()
+                    .untilAsserted(() -> {
+                        var d = k8sclient.apps().deployments().withName(deploymentName).get();
+                        assertThat(d.getMetadata().getLabels().entrySet().containsAll(labels.entrySet())).isTrue(); // additional labels should not be overwritten
+                        assertThat(d.getSpec()).isEqualTo(origSpecs); // specs should be reconciled back to original values
+                    });
+        } catch (Exception e) {
+            savePodLogs();
+            throw e;
+        }
+    }
+
+    // TODO Test HasErrors condition; ReplicaFailure is not easily reproducible
+
+    private void waitForKcToFail(Keycloak kc) {
+        Awaitility.await()
+                .ignoreExceptions()
+                .pollDelay(Duration.ofSeconds(60)) // just to be sure the cluster had enough time to deploy KC
+                .untilAsserted(() -> {
+                    var currentKc = k8sclient.resources(Keycloak.class).withName(kc.getMetadata().getName()).get();
+                    CRAssert.assertKeycloakStatusCondition(currentKc, KeycloakStatusCondition.READY, false);
+                    // we currently don't detect ImagePullBackOff, pod, config, ... errors in the operator
+                    // TODO https://github.com/keycloak/keycloak/issues/10099
+                    // CRAssert.assertKeycloakStatusCondition(currentKc, KeycloakStatusCondition.HAS_ERRORS, true);
```

Comment:
Please remove commented code, we already have an issue tracking the problem.

Reply:
👍